emperor/support_files/js/view-controllers.js

```diff
@@ -0,0 +1,210 @@
+ *
+ * @param {Node} container, Container node to create the controller in.
+ * @param {String} attrName, Attribute name to control, can be one of 'color',
+ * 'opacity' or 'visibility'.
+ * @param {Float} width, Width in pixels for the controller;
+ * @param {Float} height, Height in pixels for the controller;
+ * @param {DecompositionModel} dm, Model object with the information to be
+ * represented.
+ * @param {ScenePlotView} sceneViews, SceneViews to controll; Maybe remove and
+ * add a callback. FIXME
+ * @param {Object} options, Options to set to the controller before
+ * initializaiton.
+ *
+ * A few sections of this class are based on SlickGrid's slick.grid.js
+ *
+ **/
+function AttributeViewController(container, attrName, width, height,
+                                 dm, sceneViews, options){
+  var $container, $canvas;
+  var scope = this;
+
+  this.grid = null;
+  this.attrName = attrName;
+  this.dm = dm;
+
+  this._categorySelectionCallback = null;
+  this._valueUpdatedCallback = null;
+  this._slickGridColumn = null;
+
+  // Constant for width in slick-grid
+  this.SLICK_WIDTH = 25;
+
+  // verify we have a valid container
+  $container = $(container);
+  if ($container.length < 1) {
+    throw new Error("Emperor requires a valid container, " + container +
+        " does not exist in the DOM.");
+  }
+
+  /*
+   *
+   * Setup the needed properties to properly create the interface, mainly we
+   * will be setting up the following few features:
+   *
+   * + Callback that responds to category selection change.
+   * + Callback that responds to a value being updated.
+   * + Column configuration for SlickGrid.
+   *
+   **/
+  this._attrInit = function() {
+    if (scope.attrName === 'color'){
+      scope._categorySelectionCallback = function(evt, params) {
+        var newCategory = params.selected;
+
+        // fetch the slickgrid-formatted data
+        data = sceneViews[0].decViews[0].setCategoryColors('discrete-coloring-qiime',
+                                                           newCategory);
+        scope.grid.setData(data);
+        scope.grid.invalidate();
+        scope.grid.render();
+      };
+
+      // properly respond to color changes
+      scope._valueUpdatedCallback = function(e, args) {
+        var val = args.item.category, color = args.item.color, group = [];
+
+        group = args.item.plottables;
+        sceneViews[0].decViews[0].setGroupColor(color, group);
+      };
+
+      // we use our homebrewed color editor
+      scope._slickGridColumn = {id: 'title', name: '', field: 'color',
+                                sortable: false, maxWidth: scope.SLICK_WIDTH,
+                                minWidth: scope.SLICK_WIDTH,
+                                editor: ColorEditor,
+                                formatter: ColorFormatter};
+    }
+    else if (scope.attrName === 'opacity') {
+      // TODO
+    }
+    else if (scope.attrName === 'visibility') {
+      // TODO
+    }
```

Comment:
This will have to change, since we opted for multiple subclasses instead of one giant superclass